Feature/493 per module log levels

melekes opened the pull request:
DEPENDS ON https://github.com/tendermint/tmlibs/pull/17

ebuchman commented:
Hmm. I like that the second approach doesnt change the NewNode signature. After reading the first one I thought of a `Loggers interface` that exposes `Consensus() log.Logger` etc. so we can just pass a `loggers` into the NewNode and call the respective method to get each module logger. Then we shouldn't nee to modify tmlibs, and we'd have a unified way to init all the loggers for a tendermint node, but we'd still be changing the NewNode sig (though the number of args wouldnt change, and we'd just be replacing a `logger` with a `loggers`). But your second approach is probably better, unless I'm missing something ...

ebuchman commented on cmd/tendermint/commands/flags/log_level.go:
```diff
@@ -0,0 +1,66 @@
+	moduleToLogLevelMap = make(map[string]string)
+
+	if IsLogLevelSimple(l) {
+		if l != "info" && l != "debug" && l != "error" && l != "none" {
```
probably worth moving this check to tmlibs

ebuchman commented on cmd/tendermint/commands/run_node.go:
```diff
@@ -79,8 +81,45 @@ func runNode(cmd *cobra.Command, args []string) error {
 		}
 	}
 
+	moduleToLogLevelMap, err := tmflags.ParseLogLevel(config.LogLevel)
```
shouldn't we just put all this logic in root? `tendermint node` is prob the only command that really needs it for now, but down the road we may forget to enable if there are other commands that could benefit from fine grained logger control ...

ebuchman commented on cmd/tendermint/commands/run_node.go:
```diff
@@ -79,8 +81,45 @@ func runNode(cmd *cobra.Command, args []string) error {
+		if module == "*" {
+			switch level {
+			case "debug":
+				options = append(options, log.AllowDebug())
```
can't we do these as `AllowXxxWith` with no args ? and then only need the switch statement once ?

melekes replied:
can't we do these as AllowXxxWith with no args ?

probably not a good idea... from the user perspective `AllowInfoWith()` is like calling `SetLevel()` without passing an "info" as an arg

ebuchman commented on node/node.go:
```diff
@@ -37,9 +37,10 @@ type Node struct {
+	config              *cfg.Config
+	genesisDoc          *types.GenesisDoc    // initial validator set
+	privValidator       *types.PrivValidator // local node's validator key
+	moduleToLogLevelMap map[string]string    // logging levels per module
```
dont think this is needed now

ebuchman commented on node/node_test.go:
```diff
@@ -12,7 +12,7 @@ func TestNodeStartStop(t *testing.T) {
 
 	// Create & start node
-	n := NewNodeDefault(config, log.TestingLogger())
+	n := NewNodeDefault(config, log.TestingLogger(), make(map[string]string))
```
nope

ebuchman commented on rpc/test/helpers.go:
```diff
@@ -104,11 +104,10 @@ func NewTendermint(app abci.Application) *nm.Node {
 	privValidator := types.LoadOrGenPrivValidator(privValidatorFile, logger)
 	papp := proxy.NewLocalClientCreator(app)
-	node := nm.NewNode(config, privValidator, papp, logger)
+	node := nm.NewNode(config, privValidator, papp, logger, map[string]string{"*": "error"})
```
also nope

ethanfrey commented:
I love this work.  So nice and clean code....  I would like to move the commands/flags directory into tmlibs, so I can use this same logger parsing logic in the light-client, basecoin, etc....

I will make a few comments on the code to make it more reusable for my usecase.

ethanfrey commented on cmd/tendermint/commands/flags/log_level.go:
```diff
@@ -0,0 +1,85 @@
+)
+
+// IsLogLevelSimple returns true if log level is a single word ("info", ...).
+func IsLogLevelSimple(l string) bool {
```
Can we just expose one method that Parses simple or complex log levels?

We can keep these two, just add one that does the check for simple, like you do in root.go

ethanfrey commented on cmd/tendermint/commands/flags/log_level_test.go:
```diff
@@ -0,0 +1,42 @@
+	}
+}
+
+func TestParseComplexLogLevel(t *testing.T) {
```
Can you add a test that they work properly? Not just that they parse?

You have some nice examples of how these log levels are added... you could create a logger writing to a bytes.Buffer, then parse a log level, and send a message that should either be blocked or added, then check if it something was written to the buffer or not.

ethanfrey commented on cmd/tendermint/commands/root.go:
```diff
@@ -26,9 +28,21 @@ var RootCmd = &cobra.Command{
-		logger, err = log.NewFilterByLevel(logger, config.LogLevel)
-		if err != nil {
-			return err
+		if tmflags.IsLogLevelSimple(config.LogLevel) {
```
Can you move this logic into flags/log_level.go?

ethanfrey commented on cmd/tendermint/commands/run_node.go:
```diff
@@ -79,6 +80,14 @@ func runNode(cmd *cobra.Command, args []string) error {
 		}
 	}
 
+	if !tmflags.IsLogLevelSimple(config.LogLevel) {
```
why is this in run_node.go rather than integrated in root.go with the other parse logic?

melekes commented:
@ethanfrey addressed your comments in https://github.com/tendermint/tendermint/pull/501/commits/fb0df75de097c2b92b4dfb5396db36e02e526bbe
code looks even better now, thanks!

ebuchman commented on node/node.go:
```diff
@@ -128,7 +129,7 @@ func NewNode(config *cfg.Config, privValidator *types.PrivValidator, clientCreat
 
 	// Make MempoolReactor
-	mempoolLogger := logger.With("module", "consensus")
+	mempoolLogger := logger.With("module", "mempool")
```
lol exactly what  I was talking about today
